### PR TITLE
[special] migrate log_softmax

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference: #50345 

Rendered Docs: https://14335157-65600975-gh.circle-artifacts.com/0/docs/special.html#torch.special.log_softmax

### CODE REVIEW

File: torch/special/__init__.py [hunk 1]
```diff
@@ -504,3 +504,29 @@
 
 Alias for :func:`torch.round`.
 """)
+
+log_softmax = _add_docstr(_special.special_log_softmax,
+                          r"""
+Computes softmax followed by a logarithm.
```
Comment (mruberry) [inline torch/special/__init__.py:510]: This is an alias for https://pytorch.org/docs/master/generated/torch.nn.functional.log_softmax.html?highlight=log_softmax#torch.nn.functional.log_softmax, right?

We should probably update the OpInfo to reflect that this operation is canonically `torch.nn.functional.log_softmax`, too.

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -3597,13 +3597,13 @@ def sample_inputs_log_softmax(op_info, device, dtype, requires_grad, with_dtype=
     make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
 
     if with_dtype:
-        cases = (((S, S, S), (1, torch.float64)),)
+        cases = (((S, S, S), (1, ), {'dtype': torch.float64}),)
```
Comment (kshitij12345) [inline torch/testing/_internal/common_methods_invocations.py:3600]: Had to make `dtype` as kwarg because `nn.functional.log_softmax` has a positional argument `_stacklevel` before `dtype`.

https://pytorch.org/docs/stable/generated/torch.nn.functional.log_softmax.html#torch.nn.functional.log_softmax

File: torch/special/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/special/__init__.py:510]: Should these docs just point to the docs for torch.nn.functional.log_softmax and say this is an alias for it? 

Or is that not the case because the signatures are different?

Comment (kshitij12345) [reply torch/special/__init__.py:510]: Yeah difference in signature leads to different behaviour so.
```python
>>> import torch
>>> t = torch.ones(2,2)
>>> torch.log_softmax(t, 0, torch.float64)
tensor([[-0.6931, -0.6931],
        [-0.6931, -0.6931]], dtype=torch.float64)
>>> torch.nn.functional.log_softmax(t, 0, torch.float64)
tensor([[-0.6931, -0.6931],
        [-0.6931, -0.6931]]) # Not torch.float64
```

**Comment by kshitij12345** [comment]
@mruberry can you please check this once more to see if it is good to go :). Thanks!

File: torch/special/__init__.py [hunk 3]
```diff
@@ -505,6 +505,32 @@
...
+
+.. math::
+    \text{log\_softmax}(x_{i}) = \log\left(\frac{\exp(x_i) }{ \sum_j \exp(x_j)} \right)
+""" + r"""
+
+Args:
+    input (Tensor): input
+    dim (int): A dimension along which log_softmax will be computed.
+    dtype (:class:`torch.dtype`, optional): the desired data type of returned tensor.
+        If specified, the input tensor is casted to :attr:`dtype` before the operation
```
Comment (mruberry) [inline torch/special/__init__.py:524]: "casted" -> "cast"

File: torch/special/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/special/__init__.py:510]: That seems like a bug in nn.functional.log_softmax. Would you file an issue, @kshitij12345?

File: torch/special/__init__.py [hunk 4]
```diff
@@ -505,6 +505,32 @@
 Alias for :func:`torch.round`.
 """)
 
+log_softmax = _add_docstr(_special.special_log_softmax,
+                          r"""
+Computes softmax followed by a logarithm.
+
+While mathematically equivalent to log(softmax(x)), doing these two
+operations separately is slower, and numerically unstable. This function
```
Comment (mruberry) [inline torch/special/__init__.py:513]: no comma needed on this line

File: torch/special/__init__.py [hunk 5]
```diff
@@ -505,6 +505,32 @@
...
+uses an alternative formulation to compute the output and gradient correctly.
```
Comment (mruberry) [inline torch/special/__init__.py:514]: I think this sentence can be removed? It's not clear what this is an alternative to

File: torch/special/__init__.py [hunk 6]
```diff
@@ -505,6 +505,32 @@
 Alias for :func:`torch.round`.
 """)
 
+log_softmax = _add_docstr(_special.special_log_softmax,
```
Comment (mruberry) [inline torch/special/__init__.py:508]: This needs the signature to start the docstring, right?

File: torch/special/__init__.py [hunk 7]
```diff
@@ -505,6 +505,32 @@
...
+operations separately is slower, and numerically unstable. This function
+uses an alternative formulation to compute the output and gradient correctly.
+
+.. math::
+    \text{log\_softmax}(x_{i}) = \log\left(\frac{\exp(x_i) }{ \sum_j \exp(x_j)} \right)
+""" + r"""
+
+Args:
+    input (Tensor): input
+    dim (int): A dimension along which log_softmax will be computed.
```
Comment (mruberry) [inline torch/special/__init__.py:522]: What happens when dim is None vs. what happens when a dimension is provided?

**Review by mruberry — COMMENTED** [review]
I like the approach this takes and I agree that nn.functional.log_softmax isn't actually an alias for torch.log_softmax because of its stacklevel argument (although I think pretending it is in the OpInfo is also fine). 

I made a few comments inline about improving the docs, however

File: aten/src/ATen/native/native_functions.yaml [hunk 8]
```diff
@@ -9833,6 +9833,10 @@
   python_module: special
   variants: function
 
+- func: special_log_softmax(Tensor self, int dim, ScalarType? dtype=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:9836]: Should we make dim and dtype kwarg-only?

Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:9836]: Making `dtype` kwarg-only makes sense. 
However `dim` is a required argument so I don't think it should be kwarg-only

File: torch/special/__init__.py (hunk 1 above)
Comment (kshitij12345) [reply torch/special/__init__.py:510]: Worth an issue?

File: torch/special/__init__.py (hunk 3 above)
Comment (kshitij12345) [reply torch/special/__init__.py]: Have also updated the original docs from `nn.functional`.

File: torch/special/__init__.py (hunk 4 above)
Comment (kshitij12345) [reply torch/special/__init__.py]: Have also updated the original docs from `nn.functional`.

File: torch/special/__init__.py (hunk 7 above)
Comment (kshitij12345) [reply torch/special/__init__.py:523]: `dim` is a mandatory argument so can't be None.

File: aten/src/ATen/native/native_functions.yaml (hunk 8 above)
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml]: Have updated the signature to `special_log_softmax(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor`

**Comment by kshitij12345** [comment]
@mruberry gentle ping :)

failures look irrelevant

File: tools/autograd/gen_annotated_fn_args.py [hunk 9]
```diff
@@ -25,14 +25,17 @@
...
+    is_py_nn_function, is_py_linalg_function, is_py_variable_method, is_py_special_function, \
+    is_py_fft_function
 
 def gen_annotated(native_yaml_path: str, out: str, autograd_dir: str) -> None:
     native_functions = parse_native_yaml(native_yaml_path).native_functions
     mappings = (
         (is_py_torch_function, 'torch._C._VariableFunctions'),
         (is_py_nn_function, 'torch._C._nn'),
         (is_py_linalg_function, 'torch._C._linalg'),
+        (is_py_special_function, 'torch._C._special'),
```
Comment (mruberry) [inline tools/autograd/gen_annotated_fn_args.py:37]: Nice fix

File: torch/nn/functional.py [hunk 10]
```diff
@@ -1757,7 +1757,7 @@ def log_softmax(input: Tensor, dim: Optional[int] = None, _stacklevel: int = 3,
         input (Tensor): input
         dim (int): A dimension along which log_softmax will be computed.
         dtype (:class:`torch.dtype`, optional): the desired data type of returned tensor.
-          If specified, the input tensor is casted to :attr:`dtype` before the operation
+          If specified, the input tensor is cast to :attr:`dtype` before the operation
```
Comment (mruberry) [inline torch/nn/functional.py:1760]: Thank you for these fixes

File: torch/special/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/special/__init__.py:511]: Good idea -- might as well

File: torch/testing/_internal/common_methods_invocations.py [hunk 11]
```diff
@@ -7622,9 +7623,27 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
     OpInfo(
         'log_softmax',
         variant_test_name='dtype',
+        aliases=('special.log_softmax', 'nn.functional.log_softmax'),
         supports_out=False,
         dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
         sample_inputs_func=partial(sample_inputs_log_softmax, with_dtype=True),
+        skips=(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7630]: Great note explaining the skip

**Review by mruberry — APPROVED** [review]
Nice work, @kshitij12345

**Comment by mruberry** [comment]
Unfortunately the internal tool is struggling to rebase this, would you merge this with the latest commit @kshitij12345?
